### PR TITLE
Gradle for Prototypes

### LINKED ISSUE #337: Gradle for Prototypes
using gradle subprojects feature, prototypes can easily be built (and especially: tested) together with the main project with very little need for configuration.

A standard way to integrate Beagle into prototypes should be created and documented. This currently affects #331, #281 and #321.

### LINKED ISSUE #349: build.gradle needed for Prototypes
The current build.gradle does not work for prototypes anymore as they need another checkstyle file.

### PR DESCRIPTION
**Pull request by jGleitz**
**ATTENTION:** This PR relies on #353 and #348! Do not merge it before them. Reviewing might also be a lot easier after they have been merged.

Automatically adds all prototypes (folders in the `prototypes` folder) as gradle subprojects. These subprojects get standard tasks (java, checkstyle, etc.) configured from the parent project. Thus, prototypes that do not have any special build logic do not nedd to do anything to get built and checked.

fixes #349
fixes #337 

[Review on Reviewable](https://reviewable.io/reviews/beagle-pse/beagle/354)

### CODE REVIEW

**Comment by jGleitz** [comment]
-blocked +(needs review) 

---

Review status: 0 of 9 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-:-K6yaWYWim0Vh06jVhs5:1652064939)*

File: prototypes/README.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Prototypes
+Projects that implement a certain functionality. Instead of trying to fit into the rest of the code or covering every corner case, prototypes serve as a proof of concept, often showing how to use a certain technique.
+
+## Creating a prototype
+A prototype is simply a project on its own, which lives in a subfolder of this folder. It should have its own Eclipse project and stick to the [default layout structure](https://docs.gradle.org/current/userguide/java_plugin.html#N14F8E). Any functionality that is required by a prototype but already realised in the production code should be copied from there.
```
Comment (m1cm1c) [inline prototypes/README.md]: 'realised' → 'implemented'

File: prototypes/README.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Prototypes
+Projects that implement a certain functionality. Instead of trying to fit into the rest of the code or covering every corner case, prototypes serve as a proof of concept, often showing how to use a certain technique.
+
+## Creating a prototype
```
Comment (m1cm1c) [inline prototypes/README.md]: Headline style.

File: prototypes/README.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+A prototype is simply a project on its own, which lives in a subfolder of this folder. It should have its own Eclipse project and stick to the [default layout structure](https://docs.gradle.org/current/userguide/java_plugin.html#N14F8E). Any functionality that is required by a prototype but already realised in the production code should be copied from there.
+
+## Building a prototype
```
Comment (m1cm1c) [inline prototypes/README.md]: Headline style.

**Comment by m1cm1c** [comment]
Review status: 0 of 9 files reviewed at latest revision, 1 unresolved discussion.

---

*[build.gradle, line 6 \[r10\]](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-K6ygM2owkEgM6GpiHgX:-K6ygM2owkEgM6GpiHgY:1679015210) ([raw file](https://github.com/beagle-pse/beagle/blob/c547fd480c11a2f7066fb61ec036a19f09e265e4/build.gradle#L6)):*
What does this line do? I ask because I can't figure out how to use gradle to build a prototype. Where will the built files be found?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/beagle-pse/beagle/354)*

**Comment by m1cm1c** [comment]
Reviewed 9 of 54 files at r9.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[buildSrc/src/tasks/gradle/20 qc.gradle, line 91 \[r10\]](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-K6yhD3oCcbTwS7_BXVM:-K6yhD3pa8PbSP8Uesxb:2000062554) ([raw file](https://github.com/beagle-pse/beagle/blob/c547fd480c11a2f7066fb61ec036a19f09e265e4/buildSrc/src/tasks/gradle/20 qc.gradle#L91)):*
The file is called `Checkstyle Prototype.xml` in branch master and I don't see and change of it's name, so I suppose this is a mistake.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/beagle-pse/beagle/354)*

**Comment by m1cm1c** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[buildSrc/src/tasks/gradle/20 qc.gradle, line 91 \[r10\]](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-K6yhD3oCcbTwS7_BXVM:-K6yi5uT7FreSBIjToYa:-970395213) ([raw file](https://github.com/beagle-pse/beagle/blob/c547fd480c11a2f7066fb61ec036a19f09e265e4/buildSrc/src/tasks/gradle/20 qc.gradle#L91)):*
Just to clarify because this is sometimes hard to spot: You typed a plural s in the end of the file name.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/beagle-pse/beagle/354)*

**Comment by jGleitz** [comment]
Review status: 8 of 9 files reviewed at latest revision, 2 unresolved discussions.

---

*[build.gradle, line 6 \[r10\]](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-K6ygM2owkEgM6GpiHgX:-K6yiGobU8QheA8_5jzK:1525920281) ([raw file](https://github.com/beagle-pse/beagle/blob/c547fd480c11a2f7066fb61ec036a19f09e265e4/build.gradle#L6)):*
I’ll explain it part for part:

 * `ext`: in the main script block and in tasks you can define `ext`ernal properties. These are readable and writable from other contexts. In the main script, they apply for the project, in tasks, they apply for the task.
 * `subprojects`: A collection of all subprojects
 * `findAll`: [API method](http://docs.groovy-lang.org/latest/html/groovy-jdk/java/util/Collection.html#findAll%28groovy.lang.Closure%29)
 * `{…}` groovy syntax for a closure. The project is passed as `it`, the closure returns a boolean (as expected by `findAll`)

Overall, this line simply puts all subprojects starting with `prototype` (because they’re in the `prototypes` folder) into the project property `prototypeSubprojects`.  [This comes in handy when configuring all prototypes](https://github.com/jGleitz/Beagle/blob/gradle-for-prototypes/buildSrc/src/tasks/gradle/20%20qc.gradle#L89)

@RomanLangrehr Already found a project that tries to make compiling Eclipse add-ons simple. Ask him if this is what you are trying to achieve. I will look into this as soon as I find the time.

---

*[buildSrc/src/tasks/gradle/20 qc.gradle, line 91 \[r10\]](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-K6yhD3oCcbTwS7_BXVM:-K6ykg-xotU0ImErpPWp:304464108) ([raw file](https://github.com/beagle-pse/beagle/blob/c547fd480c11a2f7066fb61ec036a19f09e265e4/buildSrc/src/tasks/gradle/20 qc.gradle#L91)):*
You’re absolutely right. Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/beagle-pse/beagle/354)*

**Comment by jGleitz** [comment]
Review status: 7 of 9 files reviewed at latest revision, 5 unresolved discussions.

---

*[prototypes/README.md, line 4 \[r10\]](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-K6sviCidsZaIyutjGzp-r10-4:-K6z6Yd0vuKhLSdtSeog:-499146009) ([raw file](https://github.com/beagle-pse/beagle/blob/c547fd480c11a2f7066fb61ec036a19f09e265e4/prototypes/README.md#L4)):*
Done.

---

*[prototypes/README.md, line 5 \[r10\]](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-K6sviCidsZaIyutjGzp-r10-5:-K6z6WCP_IaE3AJReKjy:-499146009) ([raw file](https://github.com/beagle-pse/beagle/blob/c547fd480c11a2f7066fb61ec036a19f09e265e4/prototypes/README.md#L5)):*
Done.

---

*[prototypes/README.md, line 7 \[r10\]](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-K6sviCidsZaIyutjGzp-r10-7:-K6z6WX5KRFLhZVhmnrl:-499146009) ([raw file](https://github.com/beagle-pse/beagle/blob/c547fd480c11a2f7066fb61ec036a19f09e265e4/prototypes/README.md#L7)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/beagle-pse/beagle/354)*

**Comment by m1cm1c** [comment]
Reviewed 1 of 1 files at r11, 1 of 1 files at r12.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[build.gradle, line 6 \[r10\]](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-K6ygM2owkEgM6GpiHgX:-K6z9qGY431a-FigQEA1:1741359589) ([raw file](https://github.com/beagle-pse/beagle/blob/c547fd480c11a2f7066fb61ec036a19f09e265e4/build.gradle#L6)):*
Thank you for the explanation! Should I already see the the `*.class` files of prototypes somewhere in the `build` folder? Because I don't.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/beagle-pse/beagle/354)*

**Comment by jGleitz** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[build.gradle, line 6 \[r10\]](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-K6ygM2owkEgM6GpiHgX:-K6zBGH2l2se2f1rVxDD:-1695018946) ([raw file](https://github.com/beagle-pse/beagle/blob/c547fd480c11a2f7066fb61ec036a19f09e265e4/build.gradle#L6)):*
Theoretically, you should. Practically, gradle won’t be able to compile your code because you have dependencies to Eclipse that won’t get resolved. 

And in the moment, gradle won’t compile your code at all, becaus your prototype does not stick to the standard directory structure.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/beagle-pse/beagle/354)*

**Comment by m1cm1c** [comment]
Gradle somehow spams new folders into folder `prototypes` every time I run it. (In the folders, there are only more folders, a `MANIFEST.MF`, and a jar containing only the `MANIFEST.MF`.)

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-:-K6zEi_85RXFKZLrHy8o:-1923870196)*

**Comment by m1cm1c** [comment]
https://dl.dropboxusercontent.com/u/9402513/2016-01-01%2023%3A06%20gradle%20spaming.ogv

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-:-K6zFh96gfGboft39hxN:-1009740463)*

**Comment by jGleitz** [comment]
Oh no. This seems to be caused by the spaces. I’ll look into it tomorrow. Thanks for the video!

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-:-K6zFkjull1-DMlwA0zj:1914823976)*

**Comment by jGleitz** [comment]
I was able to reproduce and fix the issue on my machine.

---

Review status: 8 of 9 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-:-K710zS3yysNG62_qH_s:18853129)*

**Comment by acBerger** [comment]
:lgtm: 

---

Reviewed 6 of 54 files at r9, 1 of 1 files at r11, 1 of 1 files at r12, 1 of 1 files at r13.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-:-K712GVCdkwwrZUtBDf2:980077423)*

**Comment by RomanLangrehr** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[prototypes/TestPrototype/.project, line 28 \[r13\]](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-K714PbpqAaAe24BML28:-K714Pbq5t6CtoWTKnDs:1383964659) ([raw file](https://github.com/beagle-pse/beagle/blob/e6183cd02439282c406bcbdc66b07303dc52137e/prototypes/TestPrototype/.project#L28)):*
I think this file is here by mistake.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/beagle-pse/beagle/354)*

**Comment by RomanLangrehr** [comment]
I tried it on Extension Point Prototype and it always says, "Up-to-date" for checkstyle. But the checkstyle part should fail.

---

Reviewed 6 of 54 files at r9, 1 of 1 files at r11, 1 of 1 files at r12, 1 of 1 files at r13.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-:-K715bQPlZ4COm5N9N9s:-1601565223)*

**Comment by jGleitz** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[prototypes/TestPrototype/.project, line 28 \[r13\]](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-K714PbpqAaAe24BML28:-K715Vpn-UKaFxPyUIlk:-983234982) ([raw file](https://github.com/beagle-pse/beagle/blob/e6183cd02439282c406bcbdc66b07303dc52137e/prototypes/TestPrototype/.project#L28)):*
Absolutely. Thanks for pointing out! I removed it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/beagle-pse/beagle/354)*

**Comment by RomanLangrehr** [comment]
Reviewed 1 of 1 files at r14.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/beagle-pse/beagle/354)*

**Comment by jGleitz** [comment]
@RomanLangrehr: that’s the same as for @m1cm1c: gradle expects you to use the standard layout structure. It ignores all `.java` files outside of that structure (which is a good thing). So for gradle, you simply have no Java files to check.

---

Reviewed 6 of 54 files at r9, 1 of 1 files at r11, 1 of 1 files at r12, 1 of 1 files at r13.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-:-K716DijhcbIbGskFYGq:783570357)*

**Comment by RomanLangrehr** [comment]
Oh, sorry.
:lgtm: 

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-:-K71HV71cQ_PkzUgcR3c:-311915155)*

**Comment by m1cm1c** [comment]
> I was able to reproduce and fix the issue on my machine.

Thank you. This fixes it.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-:-K71Jr_eD3tuXiZ2hKO8:-879436844)*

**Comment by m1cm1c** [comment]
:lgtm: 

---

Reviewed 1 of 1 files at r13.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/beagle-pse/beagle/354#-:-K71K4Na6h0IhKXU8IPU:980077423)*

**Pull request merged**